grouper/ctl/base.py

```diff
@@ -1,29 +1,43 @@
 from abc import ABCMeta, abstractmethod
 from typing import TYPE_CHECKING
 
+from grouper.graph import Graph
+from grouper.models.base.session import get_db_engine, Session
+from grouper.repositories.factory import RepositoryFactory
+from grouper.services.factory import ServiceFactory
+from grouper.settings import settings
+from grouper.usecases.factory import UseCaseFactory
+from grouper.util import get_database_url
+
 if TYPE_CHECKING:
     from argparse import _SubParsersAction, Namespace
-    from grouper.models.base.session import Session
+    from typing import Optional
 
 
 class CtlCommand(object):
-    """Implements a subcommand of grouper-ctl."""
+    """Implements a subcommand of grouper-ctl that needs a session and a graph."""
 
     __metaclass__ = ABCMeta
 
+    @staticmethod
     @abstractmethod
-    def add_parser(self, subparsers):
-        # type: (_SubParsersAction) -> None
-        """Add the parser for this subcommand."""
+    def add_parser(subparsers):
+        # type: (_SubParsersAction) -> str
+        """Add the parser for this subcommand and return the subcommand name."""
```

Comment:
This is kind of a weird contract between CtlCommand subclasses and main, but I don't see a clear way to make it better after a quick glance at the argparse API. It would be nice if the CtlCommand subclass could return a parser that main could just add to the main parser, for instance.

Reply:
I think it's okay for it to instead have to take a stub parser that it attaches to, but returning the command name is indeed odd. My factory approach would fix that by moving the knowledge of what command name is attached to which command class into the factory.